File: cpp/day9.cpp
#include <cassert>
#include <fstream>
#include <print>
#include "tools.h"

struct FileInfo {
  int val;
  int pos;
  int size;
};

using vfi = std::vector<FileInfo>;

struct ParseFile {
  vi arr;
  vfi filled;
  vfi empty;
};

auto parseFile() -> ParseFile {
  std::ifstream file("input/day9.txt");
  assert(file);
  std::string line;
  file >> line;

  vi arr;
  vfi filled, empty;

  int f_val = 0, idx = 0;
  for (size_t ind = 0; ind < line.size(); ind++) {
    int times = line[ind] - '0';
    if (ind % 2 == 0) {
      for (int _ = 0; _ < times; _++) arr.push_back(f_val);
      if (times > 0) {
        filled.emplace_back(FileInfo{f_val, idx, times});
        f_val++;
      }
    } else {
      for (int _ = 0; _ < times; _++) arr.push_back(-1);
      if (times > 0) empty.emplace_back(FileInfo{-1, idx, times});
    }
    idx += times;
  }

  return {arr, filled, empty};
}

auto solution(vi &arr) -> uint64_t {
  size_t j = arr.size() - 1;
  uint64_t ttl = 0;
  for (size_t i = 0; i < arr.size(); i++) {
    if (arr[i] == -1) {
      while (j > i && arr[j] < 0) j--;
      if (j <= i) break;
      arr[i] = arr[j];
      arr[j] = -1;
    }
    if (arr[i] >= 0) ttl += i * arr[i];
  }

  return ttl;
}

auto solution2(vfi &filled, vfi &empty) -> uint64_t {
  uint64_t ttl = 0;
  for (int i = (int) filled.size() - 1; i >= 0; i--) {
    for (size_t j = 0; j < empty.size(); j++) {
      if (filled[i].pos <= empty[j].pos) break;
      if (filled[i].size <= empty[j].size) {
        filled[i].pos = empty[j].pos;
        empty[j].size -= filled[i].size;
        empty[j].pos += filled[i].size;
        break;
      }
    }
  }

  for (const auto &i : filled) {
    for (int ind = i.pos; ind < i.pos + i.size; ind++) {
      ttl += ind * i.val;
    }
  }

  return ttl;
}

auto main() -> int {
  auto [arr, filled, empty] = parseFile();
  auto pt1 = solution(arr);
  auto pt2 = solution2(filled, empty);
  std::print("Part 1: {}\nPart 2: {}\n", pt1, pt2);
}

File: rust/day9.rs
use std::fs;
use std::path::Path;

struct FileInfo {
    val: i32,
    pos: usize,
    size: usize,
}

type VI = Vec<i32>;
type VFI = Vec<FileInfo>;

fn parse_input(path: &Path) -> (VI, VFI, VFI) {
    let input = fs::read_to_string(path)
        .expect("Failed to read file")
        .trim()
        .chars()
        .enumerate()
        .map(|(i, c)| (i, c.to_digit(10).expect("Non-digit character") as usize))
        .collect::<Vec<_>>();

    let (arr, filled, empty, _, _) = input.iter().fold(
        (Vec::new(), Vec::new(), Vec::new(), 0, 0),
        |(mut arr, mut filled, mut empty, mut f_val, mut idx), &(ind, times)| {
            if ind % 2 == 0 {
                arr.extend(std::iter::repeat(f_val).take(times));
                if times > 0 {
                    filled.push(FileInfo {
                        val: f_val,
                        pos: idx,
                        size: times,
                    });
                    f_val += 1;
                }
            } else {
                arr.extend(std::iter::repeat(-1).take(times));
                if times > 0 {
                    empty.push(FileInfo {
                        val: -1,
                        pos: idx,
                        size: times,
                    });
                }
            }
            idx += times;
            (arr, filled, empty, f_val, idx)
        },
    );

    (arr, filled, empty)
}

fn solution(arr: &mut VI) -> u64 {
    let mut ttl: u64 = 0;
    let mut j = arr.len() - 1;

    for i in 0..arr.len() {
        if arr[i] == -1 {
            while j > i && arr[j] < 0 {
                j -= 1;
            }
            if j <= i {
                break;
            }
            arr[i] = arr[j];
            arr[j] = -1;
        }
        if arr[i] >= 0 {
            ttl += i as u64 * arr[i] as u64
        }
    }

    ttl
}

fn solution_two(filled: &mut VFI, empty: &mut VFI) -> u64 {
    for i in filled.iter_mut().rev() {
        for j in empty.iter_mut() {
            if i.pos <= j.pos {
                break;
            }
            if i.size <= j.size {
                i.pos = j.pos;
                j.size -= i.size;
                j.pos += i.size;
                break;
            }
        }
    }

    filled
        .iter()
        .flat_map(|i| (i.pos..i.pos + i.size).map(move |ind| i.val as u64 * ind as u64))
        .sum()
}

fn main() {
    let (mut arr, mut filled, mut empty) = parse_input(Path::new("../inputs/day9.txt"));
    let pt1 = solution(&mut arr);
    let pt2 = solution_two(&mut filled, &mut empty);
    println!("Part 1: {pt1}\nPart 2: {pt2}")
}
